proposals/0010-vk-buffer-ref.md

```diff
@@ -57,16 +57,17 @@ This new type will have the following operations
 
 * Copy assignment and copy construction - These copy the value of the pointer from one variable to another.
 * Dereference Method - The Get() method represents the struct const lvalue reference of the pointer to which it is applied. The selection . operator can be applied to the Get() to further select a member from the referenced struct.
-* Null Pointer Method - The IsNull() method returns true if the pointer is 0, false if not.
+* A buffer pointer can be cast to a uint64_t using uint64_t(bp). If the buffer pointer is null, the uint64_t is zero.
```

Comment:
@llvm-beanz OK. Before writing up vk::pointer_cast, what did you think of the proposal from @devshgraphicsprogramming to use/introduce reinterpret_cast?

Reply:
I don’t think we should overload the meaning of `reinterpret_cast`, which has well defined meaning in C++ that this would conflict with.

Introducing a `vk::pointer_cast` would be more consistent with the casting behavior introduced by C++’s `std::*_pointer_cast` casts. It might make sense to follow that pattern more explicitly and introduce `vk::static_pointer_cast` and `vk::reinterpret_pointer_cast` as distinctly separate casts.

Reply:
@llvm-beanz Can you give more detail about how this would conflict with the current definition of reinterpret_cast?

Looking at std::*_pointer_cast, it seems these are designed to deal with std::shared_ptr, which is a smart pointer. I am concerned that vk::pointer_cast would give the implication that buffer pointer is a smart pointer. Do you think this concern is valid?

Reply:
I was just inspired by this:
```cpp
struct T
{
    float val = 4.5f;
};

int main()
{
    T t = {};
    std::cout << t.val << std::endl;
    auto* p = &t;
    std::cout << p << std::endl;
    auto a = reinterpret_cast<uint64_t>(p);
    std::cout << a << std::endl;
    std::cout << reinterpret_cast<T*>(a)->val << std::endl;
    return 0;
}
```
however I see that `reinterpret_cast` is literally "swap the type on this" not "construct a new value of different type from this"

Maybe `static_cast<T>(val)` is a better idea in this case, because it can be defined as just `T(val)` for certain type, i.e. float to int conversions work like that.

And GLSL's way of creating `uint64_t` to/from a Buffer Reference is exactly the syntax of float-int conversions, basically implying "some" marshalling/conversion as opposed to a reinterpret.

Reply:
OK. The reinterpret_cast documentation I was looking at was not that detailed. I see the problem now.

How about casting from BufferPointer to integer, pretty much for NULL pointer check. Should we use reinterpret_cast, static_cast, or go back to an IsNull() method? I am thinking IsNull().

Reply:
we need casting not only for null checks, but also for type punning

Reply:
> How about casting from BufferPointer to integer, pretty much for NULL pointer check. Should we use reinterpret_cast, static_cast, or go back to an IsNull() method? I am thinking IsNull().

Why convert to integer? Why not instead support an overridden bool cast operator or equality/inequality operators?

There is a known bug in HLSL 2021 where overridden cast operators don't get implicitly resolved, but I think that is fixable and they do get explicitly resolved.

> we need casting not only for null checks, but also for type punning

Type punning is not casting, they are different things. Type punning is the result of how C represents types (see: [6.2.6 Representation of Types](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2310.pdf)]. Whereas C++ defines it as undefined behavior to read a different member from a union than the one you assigned, C defines the layout of unions such that you kinda get what you expect (with no guard rails).

No casting occurs in type punning, it is just union access which more or less is defined as "treat this bag of bytes as some type".

Reply:
OK. An overloaded bool cast it will be.

Since so much has changed, I will create a new PR and point to this one and I will close this one.